BootloaderCorePkg/Library/SmbiosInitLib/SmbiosInitLib.c

```diff
@@ -0,0 +1,389 @@
+#include <Library/SmbiosInitLib.h>
+#include <Library/PcdLib.h>
+#include <Library/BootloaderCommonLib.h>
+#include "SmbiosTables.h"
+
+VOID    *mType127Ptr      = NULL;
+
+
+/**
+  Check if the Smbios types (including the entry point struct)
+  have crossed the statically allocated size for Smbios init
+
+**/
+EFI_STATUS
+EFIAPI
+CheckSmbiosOverflow (
+  VOID
+  )
+{
+  SMBIOS_TABLE_ENTRY_POINT      *SmbiosEntry;
+
+  SmbiosEntry = (SMBIOS_TABLE_ENTRY_POINT *) PcdGet32 (PcdSmbiosTablesBase);
+
+  if ( (UINT32)(SmbiosEntry->EntryPointLength + sizeof (UINT8) + SmbiosEntry->TableLength) > (UINT32)PcdGet16(PcdSmbiosTablesSize) ) {
+    return TRUE;
+  } else {
+    return FALSE;
+  }
+}
+
+/**
+  Return the pointer to the Smbios table spcified by 'Type'
+
+  @param[in]  Type    Smbios type requested
+
+  @retval             Pointer to the starting of the requested
+                      Smbios type
+
+**/
+VOID *
+FindSmbiosType (
+  IN  UINT8   Type
+  )
+{
+  SMBIOS_STRUCTURE              *TypeHdr;
+  SMBIOS_TABLE_ENTRY_POINT      *SmbiosEntry;
+  UINT8                         *StringPtr;
+
+  SmbiosEntry = (SMBIOS_TABLE_ENTRY_POINT *) PcdGet32 (PcdSmbiosTablesBase);
+
+  if (SmbiosEntry == NULL) {
+    return NULL;
+  }
+
+  TypeHdr = (SMBIOS_STRUCTURE *)(SmbiosEntry->TableAddress);
+  for (; TypeHdr->Type != SMBIOS_TYPE_END_OF_TABLE;) {
+    if (TypeHdr->Type == Type) {
+      return TypeHdr;
+    }
+    //
+    // Go the end of the strings to find next Type header
+    //
+    StringPtr = (UINT8 *) TypeHdr + TypeHdr->Length;
+    while ( !(StringPtr[0] == 0 && StringPtr[1] == 0) ) {
+      StringPtr++;
+    }
+    TypeHdr = (SMBIOS_STRUCTURE *)(StringPtr + 2);
+  }
+
+  return NULL;
+}
+
+/**
+  After adding new Smbios Types, we need to do the following:
+    1. Add Type 127
+    2. Adjust the TotalLength in entry point struct
+    3. Check for the table overflow
+    4. Update entry point & intermediate checksum
+
+  @param[in]  NewMaxStructSize    Maximum type size of all the types added    , if called from SmbiosInit
+                                  Type Length of the current table appended   , if called from AppendSmbiosType
+
+**/
+VOID
+FinalizeSmbios (
+  IN  UINTN   NewMaxStructSize
+  )
+{
+  SMBIOS_TABLE_TYPE127          *Type127;
+  EFI_STATUS                    Status;
+  SMBIOS_TABLE_ENTRY_POINT      *SmbiosEntry;
+
+  SmbiosEntry = (SMBIOS_TABLE_ENTRY_POINT *) PcdGet32 (PcdSmbiosTablesBase);
+
+  if (SmbiosEntry == NULL) {
+    return;
+  }
+
+  //
+  // Create Type127 and update global Type127 for the next append
+  //
+  Type127 = (SMBIOS_TABLE_TYPE127 *) (SmbiosEntry->TableAddress + SmbiosEntry->TableLength);
+  Type127->Hdr.Type = SMBIOS_TYPE_END_OF_TABLE;
+  Type127->Hdr.Length = sizeof (SMBIOS_STRUCTURE);  
+  mType127Ptr = (VOID *) Type127;
+
+  //
+  // Update the Table Length and check for overflows
+  //
+  SmbiosEntry->TableLength += Type127->Hdr.Length + 2;
+  Status = CheckSmbiosOverflow ();
```

Comment:
Can we do the check whenever we append new record type?  
We should not allow write outside of the allocated buffer.